### ISSUE TITLE
HTTP connection aborts after server load peak

### ISSUE THREAD

**Issue by jaroslawr** [haproxy/haproxy#1377]
### Detailed Description of the Problem

After a burst of high CPU usage caused by whatever reason HAProxy is sometimes left in a state where it prematurely ends HTTP connections, while HTTPS connections continue to work as usual. When it runs into this state it also sends tons of RSTs on the backend side of the connections, which have no clear reason (no timeout has expired, backend server did not do anything funny etc.). This continues until the process is restarted.

### Expected Behavior

HAProxy should recover from overload.

### Steps to Reproduce the Behavior

Overload server with a load spike. 

### What is your configuration?

```haproxy
global
        user haproxy
        group haproxy

        nbthread 64
        cpu-map auto:1/1-64 0-31 64-95

        maxconn 120000
        hard-stop-after 10s

        tune.bufsize 32768
        tune.ssl.default-dh-param 2048
        tune.ssl.cachesize 10000000
        tune.ssl.lifetime 3600

        log /dev/log local0 info

defaults
        mode http

        log global
        option httplog

        option forwardfor
        option http-buffer-request

        maxconn 110000

        retries 0
        no option redispatch

        timeout client 30000
        timeout server 500
        timeout connect 25
        timeout queue 25

frontend http
        bind *:80

        default_backend xyz

frontend https
        bind *:443 ssl crt /etc/haproxy/abcd.pem alpn h2,http/1.1

        default_backend xyz

backend xyz
        balance leastconn
        option httpchk
        http-check send meth HEAD uri /healthcheck
        default-server check inter 1000 maxqueue 1

        server xyz_101.net_1234 xyz_101.net:1234 weight 10 maxconn 10
        ...
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.3-4dd5a5a 2021/08/17 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.3.html
Running on: Linux 5.10.0-1026-oem #27-Ubuntu SMP Thu May 6 07:19:13 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=64).
Built with OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
Running on OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

**Comment by wtarreau** [comment]
The resets on the backend might be caused by an attempt to shrink the idle connections to preserve file descriptors used on the front. That's just speculation of course.

**Comment by jaroslawr** [comment]
I attach output from `show servers conn`. It looks a bit like the `need_est` counter is stuck at some valid that is out of sync with current real connection demand?

[backend_conns.txt](https://github.com/haproxy/haproxy/files/7085325/backend_conns.txt)

**Comment by wtarreau** [comment]
It could look like this but it could also be a chicken-and-egg effect: probably that the estimate is correct but due to a lack of FDs a lot of them have been forcefully closed, explaining why "cur" is much lower than the estimate. The purpose of the estimate really is to give a direction that can be followed when all conditions are fulfilled.

**Comment by jaroslawr** [comment]
@wtarreau You are right, it looks like with our workload we simply get a CPU peak every time we cross `tune.pool-high-fd-ratio`. I attach a flamegraph of how this looks like on a server. I understood this today when I ran into problems deploying a change setting global `maxconn` to more conservative values. It caused me some confusion, because we were not near any of the `maxconn` limits. I set `maxconn` on http frontend, `maxconn` on https frontend and then global `maxconn` to the sum of the two frontends limit and `sslmaxconn` to the same limit the https frontend has. I monitored all four limits and the related connection counters (maximums since start). I needed to bump maxconn way above the historic maximum of connection count to be able to stabilize everything.

Is this something expected to happen in certain workloads? What would be a reasonable way of configuring HAProxy to avoid this? I guess just bumping maxconn to twice the frontends sum would work but seems a bit hackish, as on the first glance it doesn't seem to have anything to do with backend connections. 

[haproxy_close.zip](https://github.com/haproxy/haproxy/files/7220494/haproxy_close.zip)

**Comment by wtarreau** [comment]
Thanks a lot Jaroslaw. The flamegraph looks odd, it seems that perf couldn't proprely analyse the stack and reported some garbage.

Regardless, your observation is extremely interesting, because we've already done some optimizations in the past around the locks needed for the idle connections. Here the releasing of idle connections grabs the lock during the whole operation, forcing some serialization if other threads are currently using it.

I'm attaching two patches that should significantly improve the situation (when I placed the other trylocks I imagined that we would not face circumstances where freeing connections would be expensive, and it seems I was wrong), I'm interested in your reports on them.

Regarding the tuning of maxconn/tune.pool-high-fd-ratio, there's no good value. maxconn should always reflect what you really need, and as you say, it would be hackish (and partially dangerous) to deliberately set it too high. However if you notice you're releasing and allocating connections very fast, it means you're seeing high variations on the number of FDs and that the default ratio of 25% of FDs assigned to idle conns might be too low for you, you could think about raising it. There's nothing wrong with having even 50% if needed, what matters is that you can keep most of your connections alive. But I also think that the current situation was aggravated by the lock, because when a thread is waiting, it's not finishing its other work that could result in releasing one FD. Instead of observing a smooth activity, at the microscopic time scale you'll see waves and it's possible that you're allocating and releasing large batches at once.

So in short here's what I'd suggest:
  - 1: apply the two attached patches and test again in similar conditions (they should apply on top of 2.4 though I have not tested, they only replace `HA_SPIN_LOCK()` by `if (HA_SPIN_TRYLOCK() == 0) {` and indent the rest of the block)
  - 2: if that's not sufficient, increase `tune.pool-high-fd-ratio` to what suits your needs.
[0001-MEDIUM-backend-use-a-trylock-when-trying-to-release-.patch.txt](https://github.com/haproxy/haproxy/files/7244384/0001-MEDIUM-backend-use-a-trylock-when-trying-to-release-.patch.txt)
[0002-MEDIUM-backend-use-a-trylock-when-trying-to-pick-one.patch.txt](https://github.com/haproxy/haproxy/files/7244385/0002-MEDIUM-backend-use-a-trylock-when-trying-to-pick-one.patch.txt)

**Comment by wtarreau** [comment]
@jaroslawr just making sure you got my message above in case you missed it. No rush.

**Comment by jaroslawr** [comment]
@wtarreau Thanks, I will try to test the patches on Monday. Meanwhile as a workaround I set the global maxconn much higher than the frontend maxconns and that made the load peaks during cluster restarts and during bursts of connections/s go away, which to some extent confirms the problem.

**Comment by wtarreau** [comment]
Great, that's a good indication already, indeed!

**Comment by jaroslawr** [comment]
@wtarreau I tested the patch. I ran wrk against the version with the patch and version without the patch (client on separate machine and http://xyz is a server with our normal production configuration and in addition to the test traffic receives some real traffic):

```
./wrk -d30s -c 5000 -t32 http://xyz
```

I repeated this gradually increasing `-c` from 5 000 to 10 000 in steps of 1000. The noticeable effect of the patch is something like 5% more backend connections being closed per second, but both versions seem to suffer the same effect at about the same `-c` level - it seems the bottleneck is the kernel lock in `eventpoll_release_file`, with `-c 7000` it takes at most 2% in `perf top`, with `-c 8000`, which is around the level where the proxy should cross  `pool-low-fd-ratio`/`pool-high-fd-ratio` it already starts taking 24%.

**Comment by wtarreau** [comment]
Just thinking again about this one @jaroslawr , have you tried putting `tune.idle-pool.shared off` in your global section on the test above, and if not, could you give it a try ? I suspect that with your load and number of threads, the FD migration between the various pollers might be the reason why close() is very expensive, because the kernel probably still sees it on most if not all pollers, and must loop on all of them when closing. By disabling sharing you'll have more connections established to your servers (idle conns will be per thread) and could have impacts on your servers (think that the total number of FDs you'll see will roughly be multiplied by the number of threads, so that is not to be taken lightly), but at high loads it could be helpful. With this said, maybe tuning pool-high-fd-ratio to keep enough idle conns with shared FDs would still be better since it will cause less close() and less cache thrashing.

**Comment by jaroslawr** [comment]
@wtarreau I settled on increasing `maxconn` in `global` section by a large factor to never cross `pool-low-fd-ratio*maxsock` and to limiting frontend connections much more conservatively with `maxconn` in `frontend` section. This has worked well for some weeks now and sustained some of the real production connection peaks that happen - I am a bit reluctant to try to change it because the tests themselves are a bit on the risky side of things.

**Comment by wtarreau** [comment]
I totally understand! And quite frankly, as long as you have enough RAM on your machine to deal with FDs, you're certainly better with having plenty at your disposal for this :-)

**Comment by wtarreau** [comment]
I'm now tempted to close this one as "not in haproxy" in that the kernel side is more responsible, but leaving some hints in the doc about this and the pool-fd-ratios.
